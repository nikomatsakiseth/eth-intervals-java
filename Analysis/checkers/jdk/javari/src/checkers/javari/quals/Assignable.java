package checkers.javari.quals;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import checkers.javari.JavariChecker;

import static java.lang.annotation.ElementType.*;

/**
 * Indicates that a field is assignable, even if it is inside a {@link ReadOnly}
 * instance.
 *
 * <p>
 *
 * This annotation is part of the Javari language.
 *
 * @see JavariChecker
 * @checker.framework.manual #javari-checker Javari Checker
 */
@Documented
@Retention(RetentionPolicy.RUNTIME)
@Target({FIELD})
public @interface Assignable {

}
