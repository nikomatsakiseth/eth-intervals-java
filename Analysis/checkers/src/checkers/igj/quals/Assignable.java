package checkers.igj.quals;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import static java.lang.annotation.ElementType.*;

/**
 *
 * Indicates the annotated {@code Field} may be re-assigned regardless of the
 * immutability of the enclosing class or object instance.
 *
 * @checker.framework.manual #igj-checker IGJ Checker
 */
@Documented
@Retention(RetentionPolicy.RUNTIME)
@Target( { FIELD } )
public @interface Assignable {

}
