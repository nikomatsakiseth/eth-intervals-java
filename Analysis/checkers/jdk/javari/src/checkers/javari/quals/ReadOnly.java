package checkers.javari.quals;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import checkers.javari.JavariChecker;
import checkers.quals.SubtypeOf;
import checkers.quals.TypeQualifier;
import static java.lang.annotation.ElementType.*;

/**
 * Indicates that, for the variable on which this annotation appears,
 * the object to which this variable refers will not be modified via
 * this reference, except its fields explicitly marked as
 * {@link Mutable}.
 *
 * <p>
 *
 * This annotation is part of the Javari language.
 *
 * @see Mutable
 * @see JavariChecker
 * @checker.framework.manual #javari-checker Javari Checker
 */
@Documented
@Retention(RetentionPolicy.RUNTIME)
//@Target({FIELD, LOCAL_VARIABLE, METHOD, PARAMETER, TYPE})
@TypeQualifier
@SubtypeOf({})
public @interface ReadOnly {

}
