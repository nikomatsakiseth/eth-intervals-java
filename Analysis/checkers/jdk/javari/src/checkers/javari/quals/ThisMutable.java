package checkers.javari.quals;

import java.lang.annotation.Target;

import checkers.quals.SubtypeOf;
import checkers.quals.TypeQualifier;

/**
 * An annotation used to represent a place holder immutability type, that is
 * equivalent to the ThisMutable type on the Javari typesystem.
 *
 * However, it is an implementation detail; hence, the package-scope.
 */
@TypeQualifier
@Target({}) // A programmer cannot write @ThisMutable in a program
@SubtypeOf(ReadOnly.class)
public @interface ThisMutable {

}
