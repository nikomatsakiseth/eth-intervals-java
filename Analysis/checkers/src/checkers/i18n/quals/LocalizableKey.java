package checkers.i18n.quals;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

import checkers.quals.SubtypeOf;
import checkers.quals.TypeQualifier;
import checkers.quals.Unqualified;

/**
 * Indicates that the {@code String} type has been localized and
 * formatted for the target output locale
 */
@TypeQualifier
@SubtypeOf(Unqualified.class)
@Documented
@Retention(RetentionPolicy.RUNTIME)
public @interface LocalizableKey { }
