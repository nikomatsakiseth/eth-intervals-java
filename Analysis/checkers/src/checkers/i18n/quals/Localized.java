package checkers.i18n.quals;

import static com.sun.source.tree.Tree.Kind.*;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

import checkers.quals.*;

/**
 * Indicates that the {@code String} type has been localized and
 * formatted for the target output locale
 */
@TypeQualifier
@SubtypeOf(Unqualified.class)
@ImplicitFor( trees = {
        /* All integer literals */
        INT_LITERAL,
        LONG_LITERAL,
        FLOAT_LITERAL,
        DOUBLE_LITERAL,
        BOOLEAN_LITERAL,

        //CHAR_LITERAL,
        //STRING_LITERAL,
        //NULL_LITERAL
        }
)
@Documented
@Retention(RetentionPolicy.RUNTIME)
public @interface Localized { }
