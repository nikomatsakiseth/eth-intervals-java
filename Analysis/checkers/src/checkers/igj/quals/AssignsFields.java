package checkers.igj.quals;

import static java.lang.annotation.ElementType.*;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import checkers.quals.SubtypeOf;
import checkers.quals.TypeQualifier;

/**
 *
 * Indicates that the annotated method could assigns (but not mutate) the fields
 * of {@code this} object.
 *
 * A method with an AssignsFields receiver may not use the receiver to
 * invoke other methods with mutable or immutable reciever.
 *
 * @checker.framework.manual #igj-checker IGJ Checker
 */
// TODO: Document this

@Documented
@Retention(RetentionPolicy.RUNTIME)
//@Target( { CONSTRUCTOR, METHOD })
@TypeQualifier // (for now)
@SubtypeOf( ReadOnly.class )
public @interface AssignsFields {

}
