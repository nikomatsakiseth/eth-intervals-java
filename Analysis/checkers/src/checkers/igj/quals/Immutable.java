package checkers.igj.quals;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import checkers.quals.*;

import static java.lang.annotation.ElementType.*;

/**
 * Indicates that the annotated reference is an immutable reference to an
 * immutable object.
 *
 * An Immutable object cannot be modified. Its fields may be reassigned or
 * mutated only if they are explicitly marked as Mutable or Assignable.
 *
 * @checker.framework.manual #igj-checker IGJ Checker
 */
@Documented
@Retention(RetentionPolicy.RUNTIME)
//@Target( { FIELD, LOCAL_VARIABLE, CONSTRUCTOR, METHOD, PARAMETER, TYPE })
@TypeQualifier
@SubtypeOf( ReadOnly.class )
public @interface Immutable {

}
