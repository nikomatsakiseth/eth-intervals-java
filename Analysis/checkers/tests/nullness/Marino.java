import checkers.nullness.quals.*;
// Test cases originally written by Dan Marino, UCLA, 10/8/2007.
@checkers.quals.DefaultQualifier("Nullable")
public class Marino {

    @NonNull String m_str;
    static String ms_str;
    String m_nullableStr;

    void testWhile() throws Exception {
        String s = "foo";
        while (true) {
            @NonNull String a = s;
            System.out.println("a has length: " + a.length());
            break;
        }
        int i = 1;
        while (true){

            @NonNull String a = s;  // s cannot be null here
            s = null;
            System.out.println("hi" + s.length());
            if (i > 2) break;
            a = null;
        }
        // Checker doesn't catch that m_str not initialized.
        // This is Caveat 2 in the manual, but note that it
        // is not limited to contructors.
        System.out.println("Member string has length: " + m_str.length());

        // Dereference of any static field is allowed.
        // I suppose this is a design decision
        // for practicality in interacting with libraries...?
        System.out.println("Member string has length: " + ms_str.length());
        System.out.println("Everyone should get this error: " +
                           m_nullableStr.length());

        s = null;
        @NonNull String b = "hi";
        try{
            System.out.println("b has length: " + b.length());
            methodThatThrowsEx();
            s = "bye";
        }finally{
            // Checker doesn't catch that s will be null here.
            b = s;
            System.out.println("b has length: " + b.length());
        }
    }

    void methodThatThrowsEx() throws Exception {
    throw new Exception();
    }

}
