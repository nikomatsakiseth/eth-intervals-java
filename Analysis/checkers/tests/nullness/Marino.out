#:20: (type.incompatible)   Not an error
:22: (dereference.of.nullable)
:24: (type.incompatible)
:34: (dereference.of.nullable)
:36: (dereference.of.nullable)
:46: (type.incompatible)
